File: EDUARDO/detectordepausa_rt.m
% Classe que calcula a media dos dados quando uma nova pausa é detectada
classdef detectordepausa_rt < handle
   properties
       
       fator
       U
       Tau
       Db
       Dsp
       Dsa
       Dx
       Dy
       Dz
       Dmx
       Dmy
       Dmz
       Dma
       
       
       %Filtro só pausa
       F
       FM
      
       %Filtro só ativo
       FF
       
       %Filtro alinhamento
       FA
       FMA
       FMA2
       
   end
   methods
       %Construtor inializador do objeto  
        function obj = detectordepausa_rt
        
            obj.fator = 3;
            obj.U = 0.3;
            obj.Tau = 0.05; %Valores mais baixos, decaimento mais lento
            obj.Db = 0;
            obj.Dsp = 0;
            obj.Dsa = 0;
            obj.Dx = 0;
            obj.Dy = 0;
            obj.Dz = 0;
            obj.Dmx = 0;
            obj.Dmy = 0;
            obj.Dmz = 0;
            obj.Dma = 0;
            
            
            P=13;
            H=fir1(4*P,1/P,'high');
            G=1;
            %[H,G] = cheby2(4,60,1/1000,'high');
            obj.F=FiltroIIR(H,G);
            
            HM=ones(1,50);
            HM=HM/sum(HM);
            obj.FM=FiltroIIR(HM,1);
            
            P = 20;
            H1 = fir1(4*P,1/P,'low');
            obj.FF=FiltroIIR(H1,1);
            
            
            P=7;
            H2=fir1(4*P,1/P,'low');
            G2=1;
            obj.FA=FiltroIIR(H2,G2);
            
            HMA=ones(1,30);
            HMA=HMA/sum(HMA);
            obj.FMA=FiltroIIR(HMA,1);
            
            HMA2=ones(1,10);
            HMA2=HMA2/sum(HMA2);
            obj.FMA2=FiltroIIR(HMA2,1);
            
            
        end
        
        %an: input signal  [3x1]
        %nivel_ruido: calculado nas outras funções [1x1]
        %return :::
        %Db: Pausa detectada [1] ou Pausa não detectada [0] 
        
        function [Db,Dma] = detector_pausa(obj,an,nivel_ruido)
            
          obj.Dsp = obj.detector_so_pausa(an,nivel_ruido);
          
          obj.Dsa = obj.detector_so_ativo(an,nivel_ruido);
          
          obj.Dma = obj.detector_alinhamento(an);
          
          if obj.Dsp == 1 && obj.Dsa == 0
          
              obj.Db = 1;
            
          else
              
              obj.Db = 0;
          
          end
          
          Db = obj.Db;
          Dma = obj.Dma;
          
         end
        
        
         function Dsp = detector_so_pausa(obj,a_n2,ruido)
             
             a_n3 = obj.F.filter_rt(a_n2);
             a_n4 = abs(a_n3);
             a_n5 = obj.FM.filter_rt(a_n4);
             
             
             D = a_n5<=obj.fator*ruido;
             
             
             if sum(D)==3
                
                 obj.Dsp = 1;
                 
             else
                 
                 obj.Dsp = 0;
                 
             end
             
                Dsp = obj.Dsp; 
            
         end
          
         function Dsa = detector_so_ativo(obj,a_n2,ruido)
             
             
             a_x = a_n2 - obj.FF.filter_rt(a_n2);
             a_x2 = abs(a_x);
             
             D = a_x2>=obj.fator*ruido;
             
             dx = D(1);
             dy = D(2);
             dz = D(3);
            
             if (dx==0)
                
                 dx = obj.Dx*exp(-obj.Tau);
             
             end
             
             obj.Dx = dx;

            if (dy==0)
        
                dy=obj.Dy*exp(-obj.Tau);
    
            end
            
            obj.Dy = dy;

            if (dz==0)
        
                dz=obj.Dz*exp(-obj.Tau);
    
            end   
          
            obj.Dz = dz;

            if ((obj.Dx>obj.U) || (obj.Dy>obj.U) || (obj.Dz>obj.U))
         
                obj.Dsa = 1;
         
            else
         
                obj.Dsa = 0;
         
            end
            
                Dsa = obj.Dsa;
            
         end

   
   

   function Dma = detector_alinhamento(obj,a_n2)
             
             a_dc = obj.FA.filter_rt(a_n2);
             a_x = a_n2 - a_dc;
             a_x2 = abs(a_x);
             AC = obj.FMA.filter_rt(a_x2);
             
             DC = obj.FMA2.filter_rt(a_dc);
             
             DC = abs(DC);
             %AC = abs(AC);
             
             B = DC./AC;             
             
             DM = B>1;
             
             dmx = DM(1);
             dmy = DM(2);
             dmz = DM(3);
            
             if (dmx==0)
                
                 dmx = obj.Dmx*exp(-obj.Tau);
             
             end
             
             obj.Dmx = dmx;

            if (dmy==0)
        
                dmy=obj.Dmy*exp(-obj.Tau);
    
            end
            
            obj.Dmy = dmy;

            if (dmz==0)
        
                dmz=obj.Dmz*exp(-obj.Tau);
    
            end   
          
            obj.Dmz = dmz;

            if ((obj.Dmx>obj.U) || (obj.Dmy>obj.U) || (obj.Dmz>obj.U))
         
                obj.Dma = 1;
         
            else
         
                obj.Dma = 0;
         
            end
            
            Dma = obj.Dma;
            Dma=B(2);
            
         end
         

   
end
   
end

File: EDUARDO/APPs/AppCompleto_v12_TestThread/app/src/main/java/fun/eduardo/funcoes/detectordepausa_rt.java
package fun.eduardo.funcoes;

import net.trucomanx.pdsplibj.pdsra.*;
import net.trucomanx.pdsplibj.pdsdf.*;


public class detectordepausa_rt {

	private double fator,U,Tau,Dx,Dy,Dz;

	public double Dsp,Dsa,Db;

		
	//Filtro só pausa
	
	public PdsFir Fx,Fy,Fz;

	public PdsFir FMx,FMy,FMz;


	//Filtro só ativo

	public PdsFir FFx,FFy,FFz;


	public detectordepausa_rt(String datafile_pausa, String datafile_media, String datafile_ativo){

		this.fator = 3;
		this.U = 0.3;
		this.Tau = 0.05; //Valores mais baixos, decaimento mais lento
		this.Dx = 0;
		this.Dy = 0;
		this.Dz = 0;
		this.Dsp = 0;
		this.Dsa = 0;
		this.Db = 0;

		PdsVector H = new PdsVector(datafile_pausa);
		this.Fx = new PdsFir(H);
		this.Fy = new PdsFir(H);
		this.Fz = new PdsFir(H); 

		PdsVector HM = new PdsVector(datafile_media);
		this.FMx = new PdsFir(HM);
		this.FMy = new PdsFir(HM);
		this.FMz = new PdsFir(HM); 

		PdsVector H1 = new PdsVector(datafile_ativo);
		this.FFx = new PdsFir(H1);
		this.FFy = new PdsFir(H1);
		this.FFz = new PdsFir(H1); 

	}	

	//an: input signal  [1x3]
    //nivel_ruido: calculado nas outras funções [1x1]
    //return :::
    //Db: Pausa detectada [1] ou Pausa não detectada [0] 
        



	public void detector_so_pausa(PdsMatrix a_n2, PdsMatrix ruido) 
	{

		double a_n3[],a_n4[],a_n5[],D[];

		a_n3 = new double[3];
		a_n4 = new double[3];
		a_n5 = new double[3];
		D = new double[3];

		a_n3[0] = this.Fx.EvaluateValue(a_n2.GetValue(0,0));
		a_n3[1] = this.Fy.EvaluateValue(a_n2.GetValue(0,1));
		a_n3[2] = this.Fz.EvaluateValue(a_n2.GetValue(0,2));

		a_n4[0] = Math.abs(a_n3[0]);
		a_n4[1] = Math.abs(a_n3[1]);
		a_n4[2] = Math.abs(a_n3[2]);
					
		a_n5[0] = this.FMx.EvaluateValue(a_n4[0]);
		a_n5[1] = this.FMy.EvaluateValue(a_n4[1]);
		a_n5[2] = this.FMz.EvaluateValue(a_n4[2]);

		if (a_n5[0]<=this.fator*ruido.GetValue(0,0)) {

			D[0]=1;

		} else {

			D[0]=0;

		}

		if (a_n5[1]<=this.fator*ruido.GetValue(0,0)) {

			D[1]=1;

		} else {

			D[1]=0;

		}

		if (a_n5[2]<=this.fator*ruido.GetValue(0,0)) {

			D[2]=1;

		} else {

			D[2]=0;

		}		

		
		if ((D[0]+D[1]+D[2])==3) {

			this.Dsp=1;

		} else {

			this.Dsp=0;

		}	            
                         
             
	}

	
	public void detector_so_ativo(PdsMatrix a_n2, PdsMatrix ruido) 
	{

		double a_x[],a_x2[],D2[];

		a_x = new double[3];
		a_x2 = new double[3];
		D2 = new double[3];

		a_x[0] = a_n2.GetValue(0,0) - this.FFx.EvaluateValue(a_n2.GetValue(0,0));
		a_x[1] = a_n2.GetValue(0,1) - this.FFy.EvaluateValue(a_n2.GetValue(0,1));
		a_x[2] = a_n2.GetValue(0,2) - this.FFz.EvaluateValue(a_n2.GetValue(0,2));

		a_x2[0] = Math.abs(a_x[0]);
		a_x2[1] = Math.abs(a_x[1]);
		a_x2[2] = Math.abs(a_x[2]);
		

		if (a_x2[0]>=this.fator*ruido.GetValue(0,0)) {

			D2[0]=1;

		} else {

			D2[0]=0;

		}

		if (a_x2[1]>=this.fator*ruido.GetValue(0,0)) {

			D2[1]=1;

		} else {

			D2[1]=0;

		}

		if (a_x2[2]>=this.fator*ruido.GetValue(0,0)) {

			D2[2]=1;

		} else {

			D2[2]=0;

		}		
		

		if (D2[0] == 0)
		{
			D2[0] = this.Dx*Math.exp(-this.Tau); //VERIFICAR EXPONENCIAL
		}

		this.Dx = D2[0];
        

		if (D2[1] == 0)
		{
			D2[1] = this.Dy*Math.exp(-this.Tau); //VERIFICAR EXPONENCIAL
		}

		this.Dy = D2[1];

		
		if (D2[2] == 0)
		{
			D2[2] = this.Dz*Math.exp(-this.Tau); //VERIFICAR EXPONENCIAL
		}

		this.Dz = D2[2];

		
		if ((this.Dx>this.U) || (this.Dy>this.U) || (this.Dz>this.U))
		{
			this.Dsa = 1;
		
		} else {
			
			this.Dsa = 0;

		}               

	}


   	public double detector_pausa(PdsMatrix a_n2, PdsMatrix ruido) { 

		this.detector_so_pausa(a_n2,ruido);
		this.detector_so_ativo(a_n2,ruido);  

		if (this.Dsp == 1 && this.Dsa == 0)
		{

			this.Db = 1;

		} else {

			this.Db = 0;

		}

		return(Db);
	}


	
	

}

